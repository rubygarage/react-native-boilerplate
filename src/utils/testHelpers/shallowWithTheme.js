import React from 'react';
import { ThemeProvider } from 'styled-components/native';
// eslint-disable-next-line import/no-extraneous-dependencies
import { shallow } from 'enzyme';

import mockedTheme from './mockedTheme';

export default (tree) => shallow(
  <ThemeProvider theme={mockedTheme}>
    {tree}
  </ThemeProvider>,
);
