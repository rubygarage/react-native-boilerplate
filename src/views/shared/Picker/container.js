import React, { Component } from 'react';
import { find, propEq, prop } from 'ramda';
import { withTheme } from 'styled-components/native';
import { injectIntl } from 'react-intl';
import { compose } from 'redux';
import PropTypes from 'prop-types';

import PickerComponent from './component';

class Picker extends Component {
  state = { isOpen: false }

  get label() {
    const { selectedValue, options, intl: { formatMessage } } = this.props;
    const option = find(propEq('value', selectedValue))(options);
    return prop('label', option) || formatMessage({ id: 'shared.na' });
  }

  handlePickerToggle = () => {
    this.setState((prevState) => ({ isOpen: !prevState.isOpen }));
  }

  handleValueChange = (value) => {
    const { onValueChange } = this.props;
    onValueChange(value);
    this.handlePickerToggle();
  }

  render() {
    return (
      <PickerComponent
        {...this.props}
        {...this.state}
        label={this.label}
        onPickerToggle={this.handlePickerToggle}
        onValueChange={this.handleValueChange}
      />
    );
  }
}

Picker.propTypes = {
  onValueChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
  })).isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
  selectedValue: PropTypes.string,
};

Picker.defaultProps = {
  selectedValue: undefined,
};

const enhance = compose(
  withTheme,
  injectIntl,
);

export { Picker as PickerContainer };
export default enhance(Picker);
