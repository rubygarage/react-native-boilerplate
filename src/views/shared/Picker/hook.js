import { useState, useContext } from 'react';
import * as R from 'ramda';
import { ThemeContext } from 'styled-components';
import { useIntl } from 'react-intl';

const useContainer = (props) => {
  const {
    selectedValue,
    options,
    intl: { formatMessage },
    onValueChange,
  } = props;

  const theme = useContext(ThemeContext);
  const intl = useIntl();

  const [isOpen, setIsOpen] = useState(false);

  const detectLabel = () => {
    const option = R.find(R.propEq('value', selectedValue))(options);

    return R.prop('label', option) || formatMessage({ id: 'shared.na' });
  };

  const handlePickerToggle = () => {
    setIsOpen((prevIsOpen) => !prevIsOpen);
  };

  const handleValueChange = (value) => {
    onValueChange(value);

    handlePickerToggle();
  };

  return {
    theme,
    intl,
    isOpen,
    setIsOpen,
    detectLabel,
    handlePickerToggle,
    handleValueChange,
  };
};

export default useContainer;
