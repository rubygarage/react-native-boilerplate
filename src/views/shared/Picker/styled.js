import styled from 'styled-components/native';
import { Picker } from '@react-native-community/picker';

const PickerView = styled(Picker)`
  flex: 1;
`;

const PickerValueWrap = styled.View`
  flex-direction: row;
  align-items: center;
`;

const PickerValue = styled.Text`
  font-size: 14;
  line-height: 20;
  color: ${({ theme }) => theme.colors.black};
  margin-right: 8;
`;

export { PickerView, PickerValueWrap, PickerValue };
