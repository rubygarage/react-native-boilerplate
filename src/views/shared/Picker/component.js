import React from 'react';
import { TouchableOpacity } from 'react-native';
import PropTypes from 'prop-types';

import Icon from 'views/shared/Icon';
import SimplePicker from './SimplePicker';
import * as S from './styled';

const PickerComponent = (props) => {
  const { onPickerToggle, isOpen, label, theme } = props;
  return (isOpen
    ? <SimplePicker {...props} />
    : (
      <TouchableOpacity onPress={onPickerToggle}>
        <S.PickerValueWrap>
          <S.PickerValue>
            {label}
          </S.PickerValue>
          <Icon name="down-arrow" size={20} color={theme.colors.grey500} />
        </S.PickerValueWrap>
      </TouchableOpacity>
    ));
};

PickerComponent.propTypes = {
  onPickerToggle: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  label: PropTypes.string.isRequired,
  theme: PropTypes.shape({ colors: PropTypes.shape().isRequired }).isRequired,
};

export default PickerComponent;
