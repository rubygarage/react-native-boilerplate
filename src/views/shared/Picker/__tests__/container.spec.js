import React from 'react';

import shallowWithTheme from 'utils/testHelpers/shallowWithTheme';
import diveTo from 'utils/testHelpers/diveToEnzyme';
import Picker, { PickerContainer } from '../container';

describe('Picker container', () => {
  const onValueChange = jest.fn();
  const defaultProps = {
    selectedValue: 'value1',
    options: [
      { key: 'key0', label: 'label0', value: 'value0' },
      { key: 'key1', label: 'label1', value: 'value1' }],
    onValueChange,
  };

  const wrapper = shallowWithTheme(<Picker {...defaultProps} />);
  const container = diveTo(wrapper, PickerContainer);
  const instance = container.instance();

  jest.spyOn(instance, 'handlePickerToggle');

  it('renders correctly default', () => {
    expect(container).toMatchSnapshot();
  });

  it('returns correct label', () => {
    expect(instance.label).toEqual(defaultProps.options[1].label);
  });

  it('returns correct default label', () => {
    container.setProps({ ...defaultProps, selectedValue: undefined });
    expect(instance.label).toEqual('{Translation id: shared.na}');
  });

  it('handlePickerToggle changes state', () => {
    expect(instance.state.isOpen).toBeFalsy();
    instance.handlePickerToggle();
    expect(instance.state.isOpen).toBeTruthy();
  });

  it('handleValueChange changes value and calls handlePickerToggle', () => {
    const value = 'test value';
    instance.handleValueChange(value);
    expect(instance.handlePickerToggle).toHaveBeenCalled();
    expect(onValueChange).toHaveBeenCalledWith(value);
  });
});
