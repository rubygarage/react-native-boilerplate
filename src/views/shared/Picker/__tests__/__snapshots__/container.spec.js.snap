// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picker container renders correctly default 1`] = `
<PickerComponent
  intl={
    Object {
      "formatMessage": [Function],
    }
  }
  isOpen={false}
  label="label1"
  onPickerToggle={[Function]}
  onValueChange={[Function]}
  options={
    Array [
      Object {
        "key": "key0",
        "label": "label0",
        "value": "value0",
      },
      Object {
        "key": "key1",
        "label": "label1",
        "value": "value1",
      },
    ]
  }
  selectedValue="value1"
  theme={
    Object {
      "colors": null,
      "fonts": null,
      "sizes": null,
      "type": "Mocked theme",
    }
  }
/>
`;
